monai/visualize/occlusion_sensitivity.py

```diff
@@ -142,139 +71,184 @@ class OcclusionSensitivity:
+        Args:
+            cropped_grid: subsection of the meshgrid, where each voxel refers to the coordinate of
+                the input image. The meshgrid is created by the ``OcclusionSensitivity`` class, and
+                the generation of the subset is determined by ``sliding_window_inference``.
+            nn_module: module to call on data.
+            x: the image that was originally passed into ``OcclusionSensitivity.__call__``.
+            mul: occluded region will be multiplied by this. Can be ``torch.Tensor`` or ``float``.
+            add: after multiplication, this is added to the occluded region. Can be ``torch.Tensor`` or ``float``.
+            mask_size: Size of box to be occluded, centred on the central voxel. Should be
+                a sequence, one value for each spatial dimension.
+            occ_mode: might be used to calculate ``mul`` and ``add`` on the fly.
+            activate: if ``True``, do softmax activation if num_channels > 1 else do ``sigmoid``. If ``False``, don't do any
+                activation. If ``callable``, use callable on inferred outputs.
+            module_kwargs: kwargs to be passed onto module when inferring
+        """
+        n_batch = cropped_grid.shape[0]
+        sd = cropped_grid.ndim - 2
+        # start with copies of x to infer
+        im = torch.repeat_interleave(x, n_batch, 0)
+        # replace occluded regions
+        for b, i in enumerate(cropped_grid):
+            # get coordinates of top left corner of occluded region (possible because we use meshgrid)
+            corner_coord_slices = [slice(None)] + [slice(1)] * sd
+            # starting from corner, get the slices to extract the occluded region from the image
+            slices = [slice(b, b + 1), slice(None)] + [
+                slice(int(j), int(j) + m) for j, m in zip(i[corner_coord_slices], mask_size)
+            ]
+            to_occlude = im[slices]
+            if occ_mode == "mean_patch":
+                add, mul = OcclusionSensitivity.constant_occlusion(x, to_occlude.mean().item(), mask_size)
+
+            if callable(occ_mode):
+                to_occlude = occ_mode(x, to_occlude)
+            else:
+                to_occlude = to_occlude * mul + add
+            im[slices] = to_occlude
+        # infer
+        out: torch.Tensor = nn_module(im, **module_kwargs)
+
+        # if activation is callable, call it
+        if callable(activate):
+            out = activate(out)
+        # else if True (should be boolean), sigmoid if n_chan == 1 else softmax
+        elif activate:
+            out = out.sigmoid() if x.shape[1] == 1 else out.softmax(1)
+
+        # the output will have shape [B,C] where C is number of channels output by model (inference classes)
+        # we need to return it to sliding window inference with shape [B,C,H,W,[D]], so add dims and repeat values
+        for m in mask_size:
+            out = torch.repeat_interleave(out.unsqueeze(-1), m, dim=-1)
```

Comment:
This will return the image as `BCH'W'D'`, where `H'W'D'` are the size of the mask, which is smaller than the image. If the device is big enough to handle `BCHWD`, then `BCH'W'D'` should be fine. The for loop is unfortunate but will only be run two or three times depending on the number of spatial dimensions in the image.